Support bundling babel packages and bundle the line-ending-selector package

maxbrunsfeld opened the pull request:
Refs https://github.com/atom/atom/issues/8416

maxbrunsfeld committed:
Add line-ending-selector to packageDependencies

maxbrunsfeld committed:
Compile babel files in place during compile task

kevinsawicki commented on build/lib/uses-babel.coffee:
```diff
@@ -0,0 +1,18 @@
+fs = require 'fs'
+
+BABEL_PREFIXES = [
+	"'use babel'"
```
Super minor, but this looks like a hard tab (`\t`) here

maxbrunsfeld replied:
Oh damn, thanks. I had turned of `softTabs` for testing something, and was accidentally sprinkling hard tabs everywhere.

kevinsawicki commented on build/lib/uses-babel.coffee:
```diff
@@ -0,0 +1,18 @@
+  fs.readSync(file, buffer, 0, PREFIX_LENGTH)
+  fs.closeSync(file)
+  BABEL_PREFIXES.some (prefix) ->
+    prefix.equals(buffer.slice(0, prefix.length))
```
Should this be `prefix is buffer.slice(0, prefix.length)`? is `equals` a String method? Not familiar with it.

kevinsawicki replied:
Oh, they are buffers, nevermind.

kevinsawicki commented on build/Gruntfile.coffee:
```diff
@@ -77,6 +80,11 @@ module.exports = (grunt) ->
       ext: '.js'
 
+  babelConfig =
+    options: {}
```
Should this include any of the options in the `defaultOptions` object from `src/babel.coffee`? Not sure what the default options are for source maps.

maxbrunsfeld replied:
:+1: Yeah

maxbrunsfeld replied:
Ok, so I just grabbed the data by requiring `src/babel`, which is a bit weird. Do you think I should extract those options into some separate file, or just leave it like that for now?

kevinsawicki replied:
> Do you think I should extract those options into some separate file, or just leave it like that for now?

I think leaving it for now is fine.

maxbrunsfeld committed:
convert-tabs-to-spaces

maxbrunsfeld commented:
@kevinsawicki So I'm compiling the files *in place* at the moment, so that the `main` field in the package's `package.json` doesn't have to change. Is that going to cause problems when you re-build though, because the file gets overridden?

kevinsawicki commented:
> So I'm compiling the files in place at the moment, so that the main field in the package's package.json doesn't have to change

So they are changing the files in the `node_modules` that sits inside the clone atom/atom repository?

Could a path relative to `appDir` be specified as the `dest` value when you are pushing to the babel config array?

maxbrunsfeld commented:
> Could a path relative to appDir be specified as the dest value when you are pushing to the babel config array?

:+1: Oh yeah, of course. I didn't realize that the `compile` was already writing to the app dir. For some reason I thought that happened as a separate step.

maxbrunsfeld committed:
Write babel-compiled JS files to app dir

maxbrunsfeld committed:
Honor /** use babel */ in addition to 'use babel'

maxbrunsfeld committed:
Don't use Buffer::equals

Node 0.10 doesn't have it

maxbrunsfeld committed:
Use same babel options for compiling bundled packages

maxbrunsfeld committed:
Merge remote-tracking branch 'origin/master' into mb-bundle-line-ending-selector

maxbrunsfeld committed:
Use source-map-support to handle stack trace conversion

Previously, our Error.convertStackTrace function was provided by coffeestack,
which only works for coffee-script. This adds a dependency on 'source-map-support',
which works for any source file with inline source maps.

This also refactors the code for registering our compilers (coffee-script,
typescript, and babel) so that caching logic is shared.

maxbrunsfeld committed:
Move babel config to a separate file in static directory

ssorallen commented:
`/** @babel */` seems more appropriate than using `use`.

Turning `'use babel;` into `/** use babel */` mixes two ideas. 'use' is a keyword for [JavaScript Pragmas](http://www-archive.mozilla.org/js/language/js20-2002-04/core/pragmas.html) that can be used in '`use strict;'` to enforce strict mode within a given scope. `use` in the comment alludes to some use of actual Pragmas, but this comment is ignored by the JS interpreter.

This also follows the late JSX comment `/** @jsx ... */`, which did not have the `use` keyword.

maxbrunsfeld committed:
Map line numbers in onerror handler w/ source-map-support

nathansobo commented:
I'm down with `/** @babel */`

maxbrunsfeld committed:
Don't match newlines when finding source-map data

thomasjo commented:
> I'm down with /** @babel */

:+1: from me too. This also leaves the door open for potentially allowing developers to set some parameters as well. Not that we ever need to, but it makes a lot more sense with `/** @babel ... */`.

maxbrunsfeld committed:
Replace 'use babel' -> '@babel' in comment directive

nathansobo commented:
Thanks for the suggestion @ssorallen.

maxbrunsfeld commented:
@ssorallen :+1:

Commit:
:arrow_up: grim to drop its custom CoffeeScript stack trace support

It’s interfering with our global stack trace support.

Commit:
Merge branch 'master' into mb-bundle-line-ending-selector

Commit:
:art:

Signed-off-by: Max Brunsfeld <maxbrunsfeld@gmail.com>

Commit:
Fix inline source map regex

Need to enable multi-line mode after all

Signed-off-by: Max Brunsfeld <maxbrunsfeld@gmail.com>

Commit:
Generate sourceFiles in CoffeeScript source maps correctly

maxbrunsfeld committed:
Use compile-cache in browser process (for dev mode)

maxbrunsfeld committed:
Don't use es6 features in compile-cache

APM requires it directly, and it uses an older version of node

Commit:
Merge branch 'master' into mb-bundle-line-ending-selector

maxbrunsfeld committed:
Add sourceURL comment to all compiled js files

maxbrunsfeld commented:
Ok, I think this is ready to go. @kevinsawicki can you give this some final :eyes:?

kevinsawicki commented on build/Gruntfile.coffee:
```diff
@@ -1,6 +1,9 @@
 os = require 'os'
+glob = require 'glob'
+usesBabel = require './lib/uses-babel'
+babelOptions = require('../static/babelrc.json')
```
`.json` isn't needed here

kevinsawicki commented on src/typescript.js:
```diff
@@ -0,0 +1,53 @@
+    crypto
+      .createHash('sha1')
+      .update(sourceCode, 'utf8')
+      .digest('hex') + ".js"
```
Might be nice to use consistent quotes on this line, looks like single is the pattern in this file

kevinsawicki commented on src/compile-cache.js:
```diff
@@ -0,0 +1,131 @@
+  if (extension === '.cson') {
+    CSON.readFileSync(filePath)
+  } else {
+    var compiler = COMPILERS[extension]
```
Was there a reason not to follow this pattern for `.cson` files as well?

maxbrunsfeld replied:
We had to change the code for typescript, coffee and babel to make source maps work properly. It would have been nice to refactor the CSON code to use the same pattern, but there wasn't a pressing need to touch `season`, so we left it as is for now.

kevinsawicki commented on src/compile-cache.js:
```diff
@@ -0,0 +1,131 @@
+
+    var sourceMappingURL = lastMatch[1]
+    var rawData = sourceMappingURL.slice(sourceMappingURL.indexOf(',') + 1)
+    var sourceMap = JSON.parse(new Buffer(rawData, 'base64').toString())
```
JSON.parse handles Buffers, the `toString` isn't needed.

kevinsawicki commented on src/compile-cache.js:
```diff
@@ -0,0 +1,131 @@
+    while ((match = INLINE_SOURCE_MAP_REGEXP.exec(fileData))) {
+      lastMatch = match
+    }
+    if (lastMatch == null){
```
Missing space before `{`

kevinsawicki commented on src/compile-cache.js:
```diff
@@ -0,0 +1,131 @@
+  // source-map-support module, but we've overridden it to read the javascript
+  // code from our cache directory.
+  retrieveSourceMap: function (filePath) {
+    if (!fs.isFileSync(filePath)){
```
Missing space before `{`

kevinsawicki commented on src/compile-cache.js:
```diff
@@ -0,0 +1,131 @@
+var path = require('path')
+var CSON = require('season')
+var fs = require('fs-plus')
+var _ = require('underscore-plus')
```
It looks like this file is required before the module cache is set up in `index.js`.

That means any libraries required here and downstream won't pass through it. It would be nice to address that so there aren't any performance regressions.

maxbrunsfeld replied:
:+1: Good call. It seems like this would be addressed if we did the following:

1. Defer the require of `coffee-script` like we talked about
2. Move the setup for CSON back *out* of `compile-cache.js`, and into `index.js` after the `ModuleCache` is set up
3. Don't use `_.each` in `compile-cache`

That way, no new libraries will be required before the module cache. What do you think?

kevinsawicki replied:
> What do you think?

Sounds good.

maxbrunsfeld committed:
:art:

kevinsawicki commented on src/browser/main.coffee:
```diff
@@ -77,14 +77,9 @@ setupAtomHome = ->
-  CoffeeCache.setCacheDirectory(path.join(cacheDir, 'coffee'))
-  CoffeeCache.register()
+setupCompileCache = ->
+  compileCache = require('../compile-cache')
```
Just curious, how long does this require take in the browser process?  Just want to make sure the browser process load time is minimally impacted.

maxbrunsfeld replied:
On my MBP:

```sh
node -p 't1 = new Date().getTime(); require("coffee-cash"); new Date().getTime() - t1'
24
node -p 't1 = new Date().getTime(); require("./src/compile-cache"); new Date().getTime() - t1'
38
```

So there's a 16ms increase without asar. I would expect a smaller number when using the asar bundle. In my opinion, we should try to get that down, but not by maintaining what is now a completely duplicated code-path. We should just convert the browser-side code to JS at some point soon and eliminate that whole chunk of time.

kevinsawicki replied:
>  In my opinion, we should try to get that down, but not by maintaining what is now a completely duplicated code-path.

Yeah, I'm totally on board with de-duplication but there were only a couple lines of duplication here between the browser/render processes right? Just the part that sets the cache directory path and checks for the sudo case?

A 28ms startup time increase (14ms on the browser side, 14ms presumably on the render side) does seem a bit excessive and something to possibly investigate further before merging.

maxbrunsfeld replied:
> 14ms presumably on the render side

I would expect close to zero difference on the render side, because `babel.js`, `typescript.js`, `coffee-stack`, `coffee-cash` were all previously eager-loaded by `index.js`. This adds `source-map-support` but removes 2 other npms (`coffee-stack` and `coffee-cash`).

maxbrunsfeld replied:
Profiling startup w/ timecop now...

kevinsawicki replied:
> I would expect close to zero difference on the render side, because

Ah yeah, good point

maxbrunsfeld replied:
Ok, I see identical numbers in time-cop for every parameter (`Window load`, `Shell load`, `Workspace load`...). Are you ok :ship:ing this like this?

kevinsawicki replied:
> I see identical numbers on this branch and master for every time-cop

Great to hear :dash: 

> Are you ok :ship:ing this like this?

:shipit:

kevinsawicki commented on spec/atom-reporter.coffee:
```diff
@@ -15,8 +13,8 @@ formatStackTrace = (spec, message='', stackTrace) ->
+  stackTrace = convertedLines.join('\n')
 
-  stackTrace = convertStackTrace(convertedLines.join('\n'), sourceMaps)
   lines = stackTrace.split('\n')
```
Looks like the `convertedLines` are joined and then immediately split again now on the next line.

maxbrunsfeld replied:
:see_no_evil: Thanks; I was kind of blindly hacking away there.

kevinsawicki commented on spec/compile-cache-spec.coffee:
```diff
@@ -1,39 +0,0 @@
-CompileCache = require '../src/compile-cache'
-
-describe "Compile Cache", ->
-  describe ".addPathToCache(filePath)", ->
```
Where are these covered now?

kevinsawicki commented on src/babel.js:
```diff
@@ -0,0 +1,61 @@
+var _ = require('underscore-plus')
+var crypto = require('crypto')
+var path = require('path')
+var defaultOptions = require('../static/babelrc.json')
```
It would be nice to inline this into Atom's root mega `package.json` file since all other `.json` and `.cson` files are currently placed there instead of bundled separately.

kevinsawicki replied:
After thinking about this more, it probably isn't worth the effort/complexity, feel free to disregard.

kevinsawicki commented on src/coffee-script.js:
```diff
@@ -0,0 +1,39 @@
+'use strict'
```
Are there specs for this? I would imagine something similar to the babel/typescript ones

maxbrunsfeld replied:
I figured this was tested via the fact that the test suite loaded at all. It seemed kinda weird to add a coffee-script spec that tests that another coffee-script file can be required.

kevinsawicki replied:
Yeah, would be nice though to somewhere test that cache files are written and files with the same content are only compiled once.

Basically, what coffeestack was testing now that that logic is being done here.

maxbrunsfeld replied:
Ok, I think the caching is tested through the compile-cache spec.

kevinsawicki replied:
Awesome :+1:

maxbrunsfeld committed:
:fire: Unnecessary joining and splitting

maxbrunsfeld committed:
:racehorse: Minimize dependencies loaded before the module-cache

maxbrunsfeld committed:
Defer coffee-script require, prevent it from reassigning prepareStackTrace

kevinsawicki commented on package.json:
```diff
@@ -20,9 +20,7 @@
     "babel-core": "^5.8.21",
     "bootstrap": "^3.3.4",
     "clear-cut": "^2.0.1",
-    "coffee-cash": "0.8.0",
```
This is going to require a change in https://github.com/atom/timecop/blob/master/lib/cache-panel-view.coffee to get timecop working again.

maxbrunsfeld replied:
Woah thanks for the heads up. PR for that here: https://github.com/atom/timecop/pull/13

kevinsawicki commented on src/compile-cache.js:
```diff
@@ -0,0 +1,141 @@
+      return null
+    }
+
+    var sourceCode = fs.readFileSync(filePath, 'utf8')
```
Are read errors handled internally by `source-map-support` or would a read error here cause a notification?

maxbrunsfeld replied:
Ok, I'm just gonna let it behave like there is no source map, in the event that there's an IO error retrieving the source map.

maxbrunsfeld committed:
Add CompileCache spec

maxbrunsfeld committed:
Don't make prepareStackTrace unwritable

This breaks other code (such as Grim) that may want to temporarily
use a different prepareStackTrace function.

maxbrunsfeld committed:
Catch errors when retrieving source maps

maxbrunsfeld committed:
:fire: File extension

maxbrunsfeld committed:
:shirt: Use standard style for js code

maxbrunsfeld committed:
Remove 6to5 fixtures

maxbrunsfeld committed:
Remove harmony-collections shim in Gruntfile

eslint cannot load if this shim is present.

nathansobo commented:
:zap: Nice work.

:sunrise: The era of Babel is dawning. :sunrise:

thomasjo commented:
Bloody brilliant! Thank you so much for doing this @maxbrunsfeld! 

> :sunrise: The era of Babel is dawning. :sunrise: 

Oooooh yeah :smiling_imp:

kevinsawicki commented:
Great cleanup here as well, really nice to see this whole system level up :green_heart: :blue_heart:
